### DISCUSSION TITLE
MMDetection 3.x分支有一个bug，就是训练的mask rcnn 导出模型部署小目标效果差~

### DISCUSSION THREAD

**Discussion by vansin** [open-mmlab/mmdetection#9629] [General]
[copy from wechat mmdet group3]
我发现一个MMDetection 3.x分支有一个bug，就是训练的mask rcnn 导出模型部署的时候，比较小的物体识别效果很差很差，差到没有眼看那种。
反而默认分支的MMDetection就没有这个问题。
训练精度都很好，直接用Python检测也很好，就是导出部署模型的时候就对比较小物体识别很差很差。我用一模一样的数据做了对比。

mmdeploy部署，或者自己写的libtorch都是一样

开始我是用MMDetection 默认分支，后面换成MMDetection 3.x分支，现在发现部署的时候比较小的物体识别很差很差，又换回来MMDetection 默认分支了。

MMDetection 3.x分支识别的效果真的没有眼看了

![656a295e892ead2df3572cb4b2a9ebc](https://user-images.githubusercontent.com/25839884/212331175-2e2aa479-2eae-4ea0-b625-15357cfa2eb6.jpg)
(上图为2.x版本)

![1f11c35b37b475d805d43fe36f020a0](https://user-images.githubusercontent.com/25839884/212331202-b648e1b5-65e5-461b-81df-212253c8bfb1.jpg)
（上图为 3.x 版本）

直接用Python是正常的，只要转换模型部署就出问题了。

同一台电脑，同一份数据，同一个环境

3.x最新分支，和默认最新分支

**Comment by 1095788063** [comment]
训练精度是正常。就是导出模型的时候就识别很差了。

**Reply by vansin** [reply]
方便定量描述一下吗？

**Reply by vansin** [reply]
导出成什么 onnx tensorrt 还是？

**Reply by 1095788063** [reply]
导出 tensorrt(mmdeploy)  或 libtorch(https://github.com/threeYANG/mmdetection_libtorch)

**Reply by vansin** [reply]
https://github.com/threeYANG/mmdetection_libtorch

这个库是为 2.x 写的把，不是为mmdet 3.x 写的吧，不配套

**Reply by vansin** [reply]
mmdeploy 试过吗？

**Reply by 1095788063** [reply]
mmdeploy 都试过

**Reply by 1095788063** [reply]
![image](https://user-images.githubusercontent.com/29217820/212345161-a55d5d99-d10e-41f1-8129-9efbf6f6dbee.png)
都下载了好几个版本了

**Reply by 1095788063** [reply]
![image](https://user-images.githubusercontent.com/29217820/212345303-c88dc2ed-ce99-487f-b6c2-3bb34bab5b68.png)
![image](https://user-images.githubusercontent.com/29217820/212345398-39f02491-4766-4718-a5b5-ca26f39d0112.png)

**Reply by 1095788063** [reply]
![image](https://user-images.githubusercontent.com/29217820/212345524-d0f3c65f-a3ff-4012-b30a-2053f2fcb0f5.png)

**Reply by vansin** [reply]
掉了多少点呢？用 mmdeploy 转换了之后~

**Reply by 1095788063** [reply]
没有统计过，用肉眼都看出来不正常了。

**Reply by vansin** [reply]
方便的话可以用 mmdeploy 里面的脚本测一下应该很方便~，不方便的话反馈到这也行了~

**Reply by 1095788063** [reply]
要上班回公司才可以了。

**Reply by vansin** [reply]
您方便的时候测一下~

**Reply by vansin** [reply]
主要是部署有问题，锅首先会推到 mmdeploy 那边~

**Reply by vansin** [reply]
> 没有统计过，用肉眼都看出来不正常了。

肉眼不正常，也可以贴一张图

**Reply by 1095788063** [reply]
我用这个代码也是一样有问题 libtorch(https://github.com/threeYANG/mmdetection_libtorch) 。
3.x 分支有问题，2.x 分支没有问题，这个就可以说明跟  mmdeploy没有关系，是 3.x 分支有问题了。同时 mmdeploy 也是一样有这个问题，那么说明 3.x 分支有问题。

**Reply by 1095788063** [reply]
等下周，我做一个完整的说明，举证出来 3.x 分支有问题。

**Reply by 1095788063** [reply]
> 您方便的时候测一下~

我找到原因了，是 3.x 的版本，部署的时候尺寸输入大小必须要和训练的尺寸大小一致才正常。如果部署的时候的输入尺寸大小和训练的尺寸大小不一致就会出现这个问题。同时 rtmdet-ins 我也测试也是这个问题。

**Comment by 1095788063** [comment]
3.x 训练日志：
2023/01/12 14:05:22 - mmengine - INFO - Evaluating bbox...
2023/01/12 14:05:24 - mmengine - INFO - 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.903
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.984
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.980
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.843
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.913
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.939
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.939
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.939
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.882
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.948

2023/01/12 14:05:24 - mmengine - INFO - 
计算每个类别的 Precision (AP) @[ IoU=0.50:0.95]
+------+---------------------------+---------------------------------+
| 类id | 类名称                    | Precision (AP) @[IoU=0.50:0.95] |
+------+---------------------------+---------------------------------+
| 0    | HF-4CH                    | 0.973                           |
| 1    | HF-LA                     | 0.815                           |
| 2    | HF-LV                     | 0.894                           |
| 3    | HF-RA                     | 0.78                            |
| 4    | HF-RV                     | 0.879                           |
| 5    | PP-Chest-Axial-Planes     | 0.978                           |
| 6    | PP-Chest-Axial-Planes-Cut | 0.998                           |
+------+---------------------------+---------------------------------+
2023/01/12 14:05:24 - mmengine - INFO - bbox_mAP_copypaste: 0.903 0.984 0.980 -1.000 0.843 0.913
2023/01/12 14:05:24 - mmengine - INFO - 
------------[bbox]计算验证结果耗时：1.367415189743042秒
2023/01/12 14:05:24 - mmengine - INFO - Evaluating segm...
2023/01/12 14:05:26 - mmengine - INFO - 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.881
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.983
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.970
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.773
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.905
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.914
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.914
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.914
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.830
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.927

2023/01/12 14:05:26 - mmengine - INFO - 
计算每个类别的 Precision (AP) @[ IoU=0.50:0.95]
+------+---------------------------+---------------------------------+
| 类id | 类名称                    | Precision (AP) @[IoU=0.50:0.95] |
+------+---------------------------+---------------------------------+
| 0    | HF-4CH                    | 0.976                           |
| 1    | HF-LA                     | 0.778                           |
| 2    | HF-LV                     | 0.86                            |
| 3    | HF-RA                     | 0.726                           |
| 4    | HF-RV                     | 0.851                           |
| 5    | PP-Chest-Axial-Planes     | 0.978                           |
| 6    | PP-Chest-Axial-Planes-Cut | 0.996                           |
+------+---------------------------+---------------------------------+
2023/01/12 14:05:26 - mmengine - INFO - segm_mAP_copypaste: 0.881 0.983 0.970 -1.000 0.773 0.905
2023/01/12 14:05:26 - mmengine - INFO -

**Reply by vansin** [reply]
3.x 的 Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.903 不是比 2.x 高吗？

哪里体现了小目标检测效果不好？ 都是 -1.000

**Reply by 1095788063** [reply]
python 代码正常，导出模型部署的时候就不正常了。用mmdeploy  或者 libtorch(https://github.com/threeYANG/mmdetection_libtorch) 来部署的时候就出现识别效果很差的了。

**Comment by 1095788063** [comment]
2.x 训练日志：

2023-01-13 18:12:30,644 - mmdet - INFO - Saving checkpoint at 50 epochs
2023-01-13 18:13:02,953 - mmdet - INFO - Evaluating bbox...
2023-01-13 18:13:04,278 - mmdet - INFO - 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.869
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.986
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.977
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.799
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.879
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.905
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.905
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.905
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.832
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.916

2023-01-13 18:13:04,279 - mmdet - INFO - 
计算每个类别的 Precision (AP) @[ IoU=0.50:0.95]
+---------------------------+---------------------------------+
| 类别                      | Precision (AP) @[IoU=0.50:0.95] |
+---------------------------+---------------------------------+
| HF-4CH                    | 0.954                           |
| HF-LA                     | 0.750                           |
| HF-LV                     | 0.854                           |
| HF-RA                     | 0.748                           |
| HF-RV                     | 0.825                           |
| PP-Chest-Axial-Planes     | 0.959                           |
| PP-Chest-Axial-Planes-Cut | 0.989                           |
+---------------------------+---------------------------------+
2023-01-13 18:13:04,279 - mmdet - INFO - Evaluating segm...
2023-01-13 18:13:05,849 - mmdet - INFO - 
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.850
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.986
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.971
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.726
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.871
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.886
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.886
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.886
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.781
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.900

2023-01-13 18:13:05,850 - mmdet - INFO - 
计算每个类别的 Precision (AP) @[ IoU=0.50:0.95]
+---------------------------+---------------------------------+
| 类别                      | Precision (AP) @[IoU=0.50:0.95] |
+---------------------------+---------------------------------+
| HF-4CH                    | 0.958                           |
| HF-LA                     | 0.713                           |
| HF-LV                     | 0.826                           |
| HF-RA                     | 0.701                           |
| HF-RV                     | 0.809                           |
| PP-Chest-Axial-Planes     | 0.959                           |
| PP-Chest-Axial-Planes-Cut | 0.984                           |
+---------------------------+---------------------------------+

**Comment by 1095788063** [comment]
3.x 训练验证过程图片
![val(1)_220302-34+2w-yangyongru_052](https://user-images.githubusercontent.com/29217820/212334768-19450d62-2455-43be-8277-51a5cb864323.jpg)

**Comment by 1095788063** [comment]
2.x 训练验证过程图片
![(42-35)_211028-27+4w-huyongyi_046](https://user-images.githubusercontent.com/29217820/212334911-a48c5564-f78b-4219-b878-f8795812eaf5.jpg)

**Comment by 1095788063** [comment]
3.x 和 2.x 的代码均为最新的代码

**Comment by 1095788063** [comment]
我找到原因了，是 3.x 的版本，部署的时候尺寸输入大小必须要和训练的尺寸大小一致才正常。如果部署的时候的输入尺寸大小和训练的尺寸大小不一致就会出现这个问题。同时 rtmdet-ins 我也测试也是这个问题。

**Comment by ZwwWayne** [comment]
Thanks for your patience and report. We have noticed this issue and will fix it A.S.A.P.  Please be noted that this might take some time because we are in the vacation of the Spring Festival.

**Reply by 1095788063** [reply]
好的，我测试发现 rtmdet-ins 同样也有这个问题。修改输入尺寸，就识别效果变很差。一样的问题。
